core/src/client/async_client/mod.rs

```diff
@@ -726,95 +797,111 @@ async fn background_task<S, R>(
 		}
 	}
 
-	let mut manager = RequestManager::new();
+	let SendTaskParams {
+		mut sender,
+		from_frontend,
+		close_tx,
+		manager,
+		max_buffer_capacity_per_subscription,
+		ping_interval,
+	} = params;
 
-	let backend_event = futures_util::stream::unfold(receiver, |mut receiver| async {
-		let res = receiver.receive().await;
-		Some((res, receiver))
-	});
-	let frontend = tokio_stream::wrappers::ReceiverStream::new(frontend);
+	let mut frontend = tokio_stream::wrappers::ReceiverStream::new(from_frontend);
+	let mut rx_item = frontend.next();
+	let closed = close_tx.closed();
 
-	tokio::pin!(backend_event, frontend);
+	tokio::pin!(closed);
 
-	// Place frontend and backend messages into their own select.
-	// This implies that either messages are received (both front or backend),
-	// or the submitted ping timer expires (if provided).
-	let mut message_fut = future::select(frontend.next(), backend_event.next());
-	let mut exit_or_ping_fut = future::select(on_exit, ping_fut(ping_interval));
+	let mut exit_or_ping_fut = future::select(closed, ping_fut(ping_interval));
 
-	loop {
-		match future::select(message_fut, exit_or_ping_fut).await {
```

Comment:
So this is wherre the issue popped up before then? Offhand this all looks ok because the `select` would poll send and receive to try and make progress on each regardless of whether one of them was blocked.

The way I understood it so far, there's some global lock on send or receive in soketto or something which means that this isn't actually working though, but I don't quite follow how splitting into separate tasks would suddenly allow one to progress when the other is stuck in the same way?

Reply:
The "old task would block" until the send went through and no receive would be processed.

So if it's backpressured the receive stuff won't get the change as the background task would be stuck on a send operation.

The new code doesn't hold the mutex over `send operations` just before/after or in the receive task. Otherwise, it would deadlock at the some point.

We could have a separate task for the request manager but it's bit awkward to have two different messages to accomplish that.

Reply:
No, soketto has a BiLock where the send and receiver can make progress independently on the socket.

Reply:
> The new code doesn't hold the mutex over send operations

Where abouts does the old code hold a mutex over send operations? Offhand I can't find it :)

Reply:
No, it was just the background task that got stuck because of backpressure in the old background task, nothing really "deadlocked". That mutex is only used by the "frontend/client" and the background task doesn't care about it to make progress.

For example [this](https://github.com/paritytech/jsonrpsee/blob/master/core/src/client/async_client/mod.rs#L654-#L660) got stuck when it was backpressured because it's single-threaded and background task would block on that.

Reply:
Aha perfect; thanks! That link was what I needed; I was missing that we are awaiting there :D